<a href="https://colab.research.google.com/github/tharunika-19/Machine_Learning/blob/main/iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

WHAT THIS CELL DOES:
We are importing all the tools we need before starting.

LINE BY LINE:
- import pandas as pd → pandas helps us load and work with data in table format. "pd" is just a shortcut name.
- from sklearn.datasets import load_iris → sklearn is our ML library. load_iris loads the iris flower dataset that comes built-in.
- from sklearn.model_selection import train_test_split → this tool splits our data into training set and testing set(validation set).
- from sklearn.tree import DecisionTreeClassifier → this is our ML model. It learns patterns like a flowchart of yes/no questions.
- from sklearn.metrics import accuracy_score → this checks how accurate our model's predictions are.

KEY WORDS:
- sklearn = scikit-learn, the main ML library in Python
- import = bring this tool into our notebook
- from = specifically pick one tool from a bigger library

In [2]:
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target
print(df.head())

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0  


WHAT THIS CELL DOES:
We load the iris dataset and convert it into a table (dataframe) so we can see and work with it easily.

LINE BY LINE:
- iris = load_iris() → loads the built-in iris dataset and stores it in variable called "iris"
- pd.DataFrame(iris.data, columns=iris.feature_names) → converts raw data into a proper table with column names
- df['target'] = iris.target → adds the answer column (which flower type) to our table
- print(df.head()) → shows first 5 rows of our table

KEY WORDS:
- DataFrame = a table with rows and columns, like Excel sheet
- iris.data = the input values (measurements of flowers)
- iris.feature_names = the column names (sepal length, petal width etc)
- iris.target = the answers (0, 1, 2 = three flower types)
- df = short name we gave our table, stands for dataframe
- head() = show only first 5 rows

In [3]:
X = df.drop('target', axis=1)
y = df['target']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("Training rows:", len(X_train))
print("Testing rows:", len(X_test))

Training rows: 120
Testing rows: 30


WHAT THIS CELL DOES:
We separate our data into two parts:
- Training set → model learns from this (80%)
- Testing set → we test how well model learned (20%)
This is like studying from textbook and then writing exam on unseen questions.

LINE BY LINE:
- X = df.drop('target', axis=1) → X contains all INPUT columns (measurements). We remove the answer column.
- y = df['target'] → y contains only the ANSWERS (flower types 0,1,2)
- train_test_split(X, y, test_size=0.2, random_state=42) → splits data. 20% goes to testing, 80% to training.
- print("Training rows:", len(X_train)) → shows how many rows are used for training
- print("Testing rows:", len(X_test)) → shows how many rows are used for testing

KEY WORDS:
- X = inputs (what we feed into model)
- y = output/answers (what model should predict)
- test_size=0.2 = 20% of data kept for testing
- random_state=42 = fixes the random split so we get same result every time we run
- X_train, X_test = input data split into train and test
- y_train, y_test = answer data split into train and test

In [4]:
model = DecisionTreeClassifier()
model.fit(X_train, y_train)
print("Model trained successfully!")

Model trained successfully!


WHAT THIS CELL DOES:
We create our ML model and train it on our training data.
Training means the model looks at inputs and answers and learns the pattern.

LINE BY LINE:
- model = DecisionTreeClassifier() → creates a new empty Decision Tree model
- model.fit(X_train, y_train) → trains the model. fit() means "learn from this data"
- print("Model trained successfully!") → just confirms training is done

KEY WORDS:
- DecisionTreeClassifier = type of ML model that makes decisions using yes/no questions like a flowchart
- model = variable that stores our trained model
- fit() = the function that actually trains/teaches the model
- X_train = the input data model learns from
- y_train = the correct answers model learns from

In [5]:
predictions = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, predictions))

Accuracy: 1.0


WHAT THIS CELL DOES:
We check how well our model learned by testing it on data it has never seen before.

LINE BY LINE:
- predictions = model.predict(X_test) → model looks at test inputs and makes its guesses
- accuracy_score(y_test, predictions) → compares model's guesses with actual answers and calculates accuracy
- print("Accuracy:", ...) → shows accuracy as a number between 0 and 1 (1.0 = 100% correct)

KEY WORDS:
- predict() = function to make predictions on new data
- X_test = test inputs that model has never seen
- y_test = actual correct answers for test data
- accuracy_score = calculates what % of predictions were correct
- predictions = what our model guessed

In [6]:
import pickle

with open('iris_model.pkl', 'wb') as f:
    pickle.dump(model, f)

print("Model saved successfully!")

Model saved successfully!


WHAT THIS CELL DOES:
We save our trained model as a file so we can use it later without retraining.
Just like saving a Word document after writing.

LINE BY LINE:
- import pickle → pickle is Python's tool for saving objects as files
- open('iris_model.pkl', 'wb') → opens/creates a file called iris_model.pkl. 'wb' means write in binary format
- pickle.dump(model, f) → saves our trained model into that file
- print("Model saved successfully!") → confirms saving is done

KEY WORDS:
- pickle = Python library for saving and loading objects
- .pkl = file extension for pickle files
- 'wb' = write binary (how pickle files are saved)
- dump() = the function that saves the model into the file

In [9]:
import pandas as pd

# Predict with proper column names
sample = pd.DataFrame([[3.5, 1.0, 1.9, 0.02]],
                       columns=['sepal length (cm)',
                                'sepal width (cm)',
                                'petal length (cm)',
                                'petal width (cm)'])

result = loaded_model.predict(sample)
print("Predicted flower type:", result)

Predicted flower type: [0]


WHAT THIS CELL DOES:
We load our saved model and use it to predict on completely new data.
This is exactly how backend will use your disaster model!

LINE BY LINE:
- open('iris_model.pkl', 'rb') → opens our saved file. 'rb' means read binary
- pickle.load(f) → loads the model back from the file
- pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=[...]) → creates a new data row with proper column names
- loaded_model.predict(sample) → model predicts what flower type this is
- print("Predicted flower type:", result) → shows the prediction (0, 1, or 2)

KEY WORDS:
- 'rb' = read binary (how we open pickle files to read)
- load() = function to load saved model from file
- sample = the new unknown data we want to predict on
- DataFrame = we use this to give proper column names to our input
- result = the model's prediction